model/proto/metrics/service.proto

```diff
@@ -0,0 +1,155 @@
+  google.protobuf.Duration step = 3 [
+    (gogoproto.stdduration) = true,
+    (gogoproto.nullable) = true
+  ];
+
+  // ratePer is the duration in which the per-second rate of change is calculated for a cumulative counter metric.
+  // Optional. Default = 10m.
+  google.protobuf.Duration ratePer = 4 [
+    (gogoproto.stdduration) = true,
+    (gogoproto.nullable) = true
+  ];
+
+  // spanKinds is the list of span kinds to include (logical OR) in the resulting metrics aggregation.
+  // Optional. Default = [SPAN_KIND_SERVER].
+  repeated SpanKind spanKinds = 5;
+}
+
+// GetPerServiceLatenciesRequest contains parameters for the GetPerServiceLatencies RPC call.
+message GetPerServiceLatenciesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+  // quantile is the quantile to compute from latency histogram metrics.
+  // Required.
+  double quantile = 2;
+}
+
+// GetPerServiceCallRatesRequest contains parameters for the GetPerServiceCallRates RPC call.
+message GetPerServiceCallRatesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+}
+
+// GetPerServiceErrorRatesRequest contains parameters for the GetPerServiceErrorRates RPC call.
+message GetPerServiceErrorRatesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+}
+
+// GetPerOperationLatenciesRequest contains parameters for the GetPerOperationLatencies RPC call.
+message GetPerOperationLatenciesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+  // service_name is the service name to fetch operation metrics from.
+  // Required.
+  string service_name = 2;
+  // quantile is the quantile to compute from latency histogram metrics.
+  // Required.
+  double quantile = 3;
+}
+
+// GetPerOperationCallRatesRequest contains parameters for the GetPerOperationCallRates RPC call.
+message GetPerOperationCallRatesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+  // service_name is the service name to fetch operation metrics from.
+  // Required.
+  string service_name = 2;
+}
+
+// GetPerOperationErrorRatesRequest contains parameters for the GetPerOperationErrorRates RPC call.
+message GetPerOperationErrorRatesRequest {
+  MetricsQueryBaseRequest baseRequest = 1;
+  // service_name is the service name to fetch operation metrics from.
+  // Required.
+  string service_name = 2;
+}
+
+message GetMinStepDurationRequest{};
+
+message GetMinStepDurationResponse {
+  google.protobuf.Duration minStep = 1 [
+    (gogoproto.stdduration) = true,
+    (gogoproto.nullable) = false
+  ];
+}
+
+message GetMetricsResponse {
+  repeated Metric metrics = 1 [
+    (gogoproto.nullable) = false
+  ];
+}
+
+service MetricsQueryService {
+  // GetMinStepDuration gets the min step duration supported by the backing metrics store.
+  rpc GetMinStepDuration(GetMinStepDurationRequest) returns (GetMinStepDurationResponse);
+
+  // GetPerServiceLatencies gets latency metrics grouped by service.
+  rpc GetPerServiceLatencies(GetPerServiceLatenciesRequest) returns (GetMetricsResponse);
```

Comment:
I still have a conceptual problem with this API.

First, it does not scale well. It will work fine for small shops with a dozen services. For a 100 services it becomes iffy - is it useful for a user to see 100 charts? For orgs with 1000s of services this does not work at all, it will probably choke on the data volume, and even if it doesn't, it's completely useless to return 1000s of time series to the UI.

Second, I don't think this is the right user workflow. The notion of a "service" is pretty loose, a serverless function can be a service, or an instance of ML model being served, so even small shops can easily get into a state with 100s and 1000s of "services". There's no user persona that need to see all this data all at once, nor see it ranked in a flat list, because the requirements for latency are very different for different services, it doesn't make sense to just sort and return top-K.

Why not start with having this API serve a "default dashboard" for a single service? That's a clear use case and a well understood user workflow. To extend this to multiple services requires some form of grouping so that we don't pull the data for all services at once.

Reply:
>Why not start with having this API serve a "default dashboard" for a single service?

The UI would then first get a list of services, decide which ones to place on the screen, and run queries passing the service they are interested in?

Reply:
No, I am thinking the user will pick a service, the UI won't decide by itself. It's essentially how the search works today as well, so will be consistent behavior in the UI. I don't think it fits exactly the vision that Albert had, but per my comment above I don't think that vision scales well with # of services.

Reply:
Instead of requiring the user to select a service to get useful data, we could preemptively show data about some services, even if only the 10 services with the most recent activity.

Reply:
Thanks for your feedback @jpkrohling & @yurishkuro.

> First, it does not scale well. 

I agree, this will not scale for 100s+ services, especially for latency computations which is at least an O(n^3) problem. A single service-level view would reduce this to an O(n^2) problem.
As such, I'll remove those higher-level cross-service endpoints from the API, and we'll rethink the UI design to accommodate for this new workflow, where the user will select the service to display its list of operations' metrics.

> There's no user persona that need to see all this data all at once, nor see it ranked in a flat list, because the requirements for latency are very different for different services, it doesn't make sense to just sort and return top-K.

I think there are valid use cases for  a higher-level view of services' metrics.

Some example use cases (feel free to refute if these examples seem spurious):
- Post-deployment sanity checks: As a developer or devops engineer, I want to be assured that my deployment does not negatively impact other services (esp. those that are not immediate dependencies) and therefore the business in general.
- As an engineer, I want to quickly pin-point the cause of a vague problem identified by customers (the website is slow) and avoid the back and forth conversation to find the slow service.

I agree that different services have different latency and perhaps even error profiles. Ideally, a change in error rate, latency, etc. or some sort anomaly detection would have the most value (especially for the use cases listed above), though would be more complex and so this proposal would be a more reachable stepping stone towards that direction.

Yes, there's little value in returning all services, although I thought there is some value in sorting and returning top-K, especially if it's based on a delta measurement. Again, this is compute intensive and would not be a scalable solution in this iteration.

Given the above, we will continue thinking of better ways to compute and present this higher-level service view to users in a scalable manner (e.g. post-processing a spoon-fed "view" of metrics) as we still believe it is a valuable feature and also welcome suggestions.

Reply:
> Instead of requiring the user to select a service to get useful data, we could preemptively show data about some services, even if only the 10 services with the most recent activity.

Thanks for the idea, Juraci; a related idea from my colleague is to let users choose their `k` "favourite" services (e.g. user owns the service or services are prone to issues) and only show these services.

In addition somewhere on the screen we can say something like: "looking for another service? Search for it here" with a search bar, and the result will take them into the service they looked for.

This reduces load as we query on a limited set of services' metrics and we could even enforce a hard limit of services (e.g. 20), while providing the means to still sort on the top-k services.

In future, if we can offer the ability to sort _all_ services and showing the top-k, we could use this data as a "smart suggestion" of services that are interesting, instead of always showing a view of all services sorted by top-k. So the more expensive "smart suggestion" query would be limited to when users wish to add a new "favourite" and not all the time.

Reply:
>a related idea from my colleague is to let users choose their k "favourite" services

We don't have a notion of users in Jaeger, we'd have to store their preferences in a cookie. Is that good enough?

Reply:
> We don't have a notion of users in Jaeger, we'd have to store their preferences in a cookie. Is that good enough?

Unless if I'm mistaken, cookies are better suited for cases where there are authenticated users with the web-server and there is a need for shared state between client and server.

Given there isn't a notion of (nor even a need for) users/authN afaik, I think [localStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) would be well-suited for our use case as jaeger-ui will be sending HTTP requests to jaeger-query, passing the service parameter that is stored locally on the client-side in localStorage.

Reply:
> although I thought there is some value in sorting and returning top-K

My point was that comparing latencies across services is pretty meaningless. There can always be some kind of batch processing service triggered by an RPC from a cron job, whose latency will always be high, but it doesn't mean that it's abnormal. Whereas a critical shared service (like cache) whose latency went from 1ms to 2ms won't make the top-K but it's clearly a drastic regression.

Reply:
I agree. The intention of the first set of PRs was to use this as a stepping-stone towards something more meaningful like showing the change in the metric or something more sophisticated like anomaly detection; given it's a bit more complex. Additionally, we included an "Impact" column in the [proposed UI mockup](https://github.com/jaegertracing/jaeger-ui/issues/736) that multiplies the latency with the call rate, which helps make it a bit more meaningful; though of course, a delta of the metrics would amplify the positive signal better I think.

Would you prefer that we tackle this now rather than later? That is, at least something simpler like if the response were to include a delta between the last data-point of the current time window `(now-lookback, now]` and the last data-point of the previous time window `(now-lookback*2, now-lookback]`, or something similar?

Reply:
Please let me know if there are any further questions/concerns that need addressing in this PR.

Reply:
Given the positive reaction regarding the mockup, I would go with the minimal solution that works, release it, and iterate.

Reply:
Thanks @jpkrohling :) 

I've updated the API to support metrics aggregated at a service level **and** operation level, requiring the client to explicitly pass service names at all times with the following goals in mind:
- Address the scalability concern by not returning all service metrics
- Support  the proposed "user preference" list of metrics grouped by service.
- Support per-service set of metrics grouped by operation, while giving clients the flexibility to fetch more than one service's set of metrics concurrently upfront if they so wish, though expect the more common use case to be just a single service (the single service view with multiple RED metrics per operation).
- Minimize "surface area" of the API, with little scope for submitting custom queries.
- Simplify the API by minimizing the number of RPC endpoints.